### ISSUE TITLE
3.0.0 `gapindex::get_data` function triggers space quota issues

### ISSUE THREAD

**Issue by EmilyMarkowitz-NOAA** [afsc-gap-products/gapindex#63]
### Issue Description

TLDR: The new 3.0.0 `gapindex::get_data` function triggers space quota issues. Tagging in @sowasser for awareness. 

### Steps to Reproduce

When I run the [species complex example on gapindex](https://afsc-gap-products.github.io/gapindex/articles/ex_species_complex.html), everything runs great! (yay!)

![Image](https://github.com/user-attachments/assets/7dd1c6ba-0398-4221-a9d5-cd69a029e97a)

However, in running the [download data script for the Bering Sea data report](https://github.com/afsc-gap-products/gap_bs_data_report/blob/2953b830055b1eb6a1d1391af5559af837c4cea8/code/data_dl.R#L226), pulling through the `GAP_PRODUCTS` schema, I experienced this issue: 

![Image](https://github.com/user-attachments/assets/39d3b6e8-2e88-4c07-92df-2c6575b684ae)

And when we try running this through a personal schema (here, @sowasser's personal oracle schema) the same error occurs much sooner: 

![Image](https://github.com/user-attachments/assets/b1651b7d-c587-41bb-815c-8dda660e79e1)

So I tried running the `get_data` function line by line, and got `Oracle error 1536` indicating a the space quota issue. As described in this [StackOverflow issue](https://stackoverflow.com/questions/4493485/what-is-oracle-error-ora-01536), this issue is due to: 
> "**Cause**: The space quota for the segment owner in the tablespace has been exhausted and the operation attempted the creation of a new segment extent in the tablespace."
> "**Action**: Either drop unnecessary objects in the tablespace to reclaim space or have a privileged user increase the quota on this tablespace for the segment owner."

This explains why the `GAP_PRODCUTS` and a personal schema would trigger space errors at different sections of the code - `GAP_PRODCUTS` schema has more space! 

![Image](https://github.com/user-attachments/assets/317a1757-a3d9-41ec-931a-46f78d7c9a61)

### Steps to Reproduce

Here is the excerpt of code from the [Bering Sea data report download data script](https://github.com/afsc-gap-products/gap_bs_data_report/blob/2953b830055b1eb6a1d1391af5559af837c4cea8/code/data_dl.R#L226):

```
# Load libraries and connect to Oracle
library(gapindex)
channel <- gapindex::get_connected()

# Species Covered
googledrive::drive_download(file = googledrive::as_id("10Pn3fWkB-Jjcsz4iG7UlR-LXbIVYofy1yHhKkYZhv2M"),
                            type = "csv",
                            overwrite = TRUE,
                            path = paste0(dir_out_rawdata, "/species-local-names"))

# identify which species complexes you need
report_spp <- readr::read_csv(file = paste0(dir_out_rawdata, "/species-local-names.csv"), 
                              skip = 1, 
                              show_col_types = FALSE) %>%  
  dplyr::filter(grepl(x = species_code, pattern = "c(", fixed = TRUE)) 

temp1 <- data.frame()
for (i in 1:nrow(report_spp)){
  temp2 <- eval(expr = parse(text = report_spp$species_code[i]))
  temp1 <- dplyr::bind_rows(temp1, 
                            dplyr::bind_cols(GROUP_CODE = report_spp$print_name[i], 
                                             SPECIES_CODE = eval(expr = parse(text = report_spp$species_code[i]))))
}

## Pull data. 
production_data <- gapindex::get_data(
  year_set = 1982:maxyr,
  survey_set = "EBS"
  spp_codes = temp1,
  pull_lengths = TRUE, 
  haul_type = 3, 
  abundance_haul = "Y", 
  taxonomic_source = "GAP_PRODUCTS.TAXONOMIC_CLASSIFICATION", # same thing happens with "RACEBASE.SPECIES"
  channel = channel)
```

### Suggested solution

I'm not sure exactly what the solution is, but maybe there is a way to increase the space when we run the `get_data` function? This code worked in the last version of `gapindex`, and the function doesn't look that different, so I'm not sure what changed. 

Thanks for your help, @zoyafuso-NOAA!

**Comment by zoyafuso-NOAA** [comment]
Thanks for bringing this up and it's an interesting problem. The space problem is occurring because the previous development version (now 3.0.0) of gapindex writes temporary tables to Oracle to help with queries. I recommend two workflow changes to avoid this error and possibly speed up the data pull:

1) Filter the `temp1` dataframe to only existing SPECIES_CODE values. This reduces `temp1` from  283075 records to 10347 records. 

```
# filter temp1 for exisiting species codes in GAP_PRODUCTS.TAXONOMIC_CLASSIFICATION
# note. maybe gp_taxa is already called somewhere but directly pulling here..
gp_taxa <- 
  RODBC::sqlQuery(channel = channel, 
                  query = "SELECT * FROM GAP_PRODUCTS.TAXONOMIC_CLASSIFICATION 
                           WHERE SURVEY_SPECIES = 1 ORDER BY SPECIES_CODE")
temp1 <- subset(x = temp1,
                subset = SPECIES_CODE %in% gp_taxa$SPECIES_CODE)
```

2) Have two separate data pulls. In the first data pull, you would query catch and effort data for all records in `temp1` (pull_lengths = FALSE) and then calculate CPUE and biomass only. Then, in a second data pull, query catch, effort, and size data (pull_lengths = TRUE) for just single taxa and calculate length compositions. In this second call, do you need length compositions for all survey years or just the most recent one(s)? If so, you can further speed up the call by reducing the year set, e.g., `year_set = maxyr` or `year_set = maxyr:(maxyr - 1)`. You can also replace this second call with a query to AKFIN_SIZECOMP if it's just species. 

This could speed up your data pull because size data are not being collated for taxon complexes and size data are only being pulled for species-level taxon codes. Let me know if either of these works, especially when on a personal account. I hope at least some of this helps.

**Comment by EmilyMarkowitz-NOAA** [comment]
Thanks for the quick and helpful response, @zoyafuso-NOAA !

1. Your suggestions made the run much faster (really great suggestion!) but it still buckled at the same spot (using the `GAP_PRODUCTS` Oracle schema login). 
![Image](https://github.com/user-attachments/assets/8cb51e94-d4a0-4796-a8bd-82950050fdee)

2. In troubleshooting this and looking through your suggestion, I actually realized that I don't need sizecomp data for these complexes at all... so I simply set `pull_lengths = FALSE` and it completed! 
![Image](https://github.com/user-attachments/assets/76f20be3-165f-4f6b-8292-71c675608c07)

I tried running these with my personal Oracle schema and it also worked. @sowasser can you try the script with your oracle schema and see if it will finish for you? The code is in my `em` branch: https://github.com/afsc-gap-products/gap_bs_data_report/commit/68066531cf3a5ede78f0d41b866045b9d0c8a168.

**Comment by sowasser** [comment]
I am running into the same error as before. 

![Image](https://github.com/user-attachments/assets/a2aeeef2-5c4d-475d-a436-27fdc26657b8)

**Comment by zoyafuso-NOAA** [comment]
After further discussion it seems that there are a couple of lingering issues but a way forward. 

- @sowasser's Oracle account has a small tablespace quota compared to most Oracle users at AFSC, possibly due to how the account was set up. Using GAP_PRODUCTS credentials seems to circumvent this issue.
- The next release in gapindex should involve temporary instead of permanent table creation within the intermediate steps in the get_data() function to avoid having to use permanent table space, if my understanding of Oracle tablespaces is correct.

**Comment by EmilyMarkowitz-NOAA** [comment]
Sounds like a good plan, @zoyafuso-NOAA! For clarification, I think that the most people are given @sowasser's table space when they are first given an Oracle account. I think we (my account, your account, and GAP_PRODUCTS) all have more because we requested more space, which is different from the norm, and that most users will experience this issue. Let us know if we can test anything for you!

**Comment by zoyafuso-NOAA** [comment]
I don't remember asking for an increase in my quota, but maybe I was defaulted a larger space when my account was created five years ago? I'm just noticing now that @sowasser get_data() call stops at the first query when the survey-level data are being queried, which is a very small table (on the order of dozens). So yes, if most folks have similarly low levels of space then most will also get this issue. Sophia mentioned that she received her account while she was a postdoc so maybe OFIS has different onboarding defaults for AFSC folks vs non-AFSC folks? I'll investigate offline further.  Thanks again

**Comment by zoyafuso-NOAA** [comment]
Alternatively, it could be because the recycle bin is too full. When tables are dropped, they're moved there and over time they could be accumulating enough space to trigger a space error. I'm toying around with this as well

```
RODBC::sqlQuery(channel = channel, query = "select * from USER_TS_QUOTAS;" )
RODBC::sqlQuery(channel = channel, query = "PURGE RECYCLEBIN;" )
RODBC::sqlQuery(channel = channel, query = "select * from USER_TS_QUOTAS;" )
```

Purging the recycle bin may free up some space.

**Comment by sowasser** [comment]
Thank you for this, @zoyafuso-NOAA. The `gapindex::get_data()` example from [the gapindex vignette](https://afsc-gap-products.github.io/gapindex/articles/ex_workflow.html) worked once I cleared my recycle bin. 

Unfortunately, when trying to run [the data download for the Bering Sea data report](https://github.com/afsc-gap-products/gap_bs_data_report/blob/3a5e5a0a30ea45ddd40859ed6ce36bc36285c592/code/data_dl.R#L163), I get the same error as before.

![Image](https://github.com/user-attachments/assets/e57a29bf-941d-4126-bc36-45b979a7826b)

**Comment by sowasser** [comment]
Hi @zoyafuso-NOAA,

I'm trying to pull the pollock length data so I can get age comps and I'm running into this error again.

![Image](https://github.com/user-attachments/assets/bf5eb3b6-59fb-4f9c-976d-99f990393541)

I have just reinstalled gapindex, so I'm working with version 3.0.2, and I've tried clearing my Oracle recycle bin.

![Image](https://github.com/user-attachments/assets/98946f3f-dfa3-4ba9-b541-8d9d3cd16b4b)

Do you have a sense of what else I could be doing wrong here?

Thanks!!

**Comment by zoyafuso-NOAA** [comment]
Oof, sorry about the trouble there. I'm not sure how to go about this but I can survey other folks down the hall and see what the culprit it. My guess is that you don't have enough user space but you can run the example from [the gapindex vignette](https://afsc-gap-products.github.io/gapindex/articles/ex_workflow.html) right? Can you check which Oracle Driver you are using?

```
## Restart R and make sure to have the latest gapindex version
devtools::install_github("afsc-gap-products/gapindex")
library(gapindex)

## Connect to Oracle. Make sure you are either connected to the internal NOAA
## network or VPN. 
channel <- gapindex::get_connected(check_access = F)

## Record which ODBC Driver is being used for the AFSC database
## Record user space details
RODBC::sqlQuery(channel = channel, query = "select * from USER_TS_QUOTAS;" )

gapindex_data_ebs <- gapindex::get_data(
  year_set = c(1982:2024),
  survey_set = "EBS",
  spp_codes = 21740,
  haul_type = 3,
  abundance_haul = "Y",
  pull_lengths = TRUE,
  channel = channel
)

## Did it run?
```

1) Zack Oyafuso
Using R Version 4.3.3
Oracle in OraClient12Home1
  TABLESPACE_NAME     BYTES          MAX_BYTES  BLOCKS  MAX_BLOCKS DROPPED
  USERS                          163774464  262144000   19992      32000             NO
Example query successfully ran. 

2) GAP_PRODUCTS 
Using R Version 4.3.3
Oracle in OraClient12Home1
TABLESPACE_NAME      BYTES   MAX_BYTES BLOCKS MAX_BLOCKS DROPPED
RACE_WORK 3897622528 20971520000 475784    2560000      NO
Example query successfully ran. 

3) Sophia Wasserman
Using R Version 4.3.3
[Oracle Driver specifics]
TABLESPACE_NAME      BYTES   MAX_BYTES BLOCKS MAX_BLOCKS DROPPED
USERS 0 20971520 0 2560 NO
**Example query did not successfully run**

3) Alex Dowlin
Using R Version 4.4.1
Oracle in InstantClient_19_18
TABLESPACE_NAME      BYTES   MAX_BYTES BLOCKS MAX_BLOCKS DROPPED
USERS 0 262144000 0 32000 NO
**Example query did not successfully run**

**Comment by sowasser** [comment]
Hi Zack,

Thanks again for looking into this. I have made sure that I updated gapindex. I cannot run the data pull in the vignette either. I did try my original query using the GAP Products credentials and it did work, so it looks like an issue with my Oracle account/credentials.

**Comment by zoyafuso-NOAA** [comment]
Notes to self:
- The version of the Oracle Driver does not seem to be the issue, get_data() can be called for users who have either OraClient12Home1 or InstantClient_19_18.
- The problem seems to be account-specific. For example, @AlexandraDowlin-NOAA could not run the vignette when logged into her account on her machine but the vignette can run when @zoyafuso-NOAA used his credentials. @sowasser could not run the vignette on her machine with her credentials but could run the vignette with GAP_PRODUCTS credentials.
- The issue could be traced to whether the user has write access to their account. Many of the content in the get_data() function is the writing of temporary tables to the user's schema. If they do not have write access, they cannot use the function. @AlexandraDowlin-NOAA and @sowasser currently do not have write access to their own schemata. 
- @sowasser and @AlexandraDowlin-NOAA, could you ask OFIS to update your Oracle accounts to include write access to your accounts so you can create tables and then try the vignette above again.

**Comment by sowasser** [comment]
Hi @zoyafuso-NOAA and @AlexandraDowlin-NOAA, I was able to get my permissions in Oracle changed and I can now write temporary tables. The gapindex vignette runs (see 2nd call in the screenshot below), but the data pull I need does not. 

![Image](https://github.com/user-attachments/assets/1a518fa1-b5be-4fc5-9bb4-36da092c31bd)

I am thinking this is the space issue, now. I did try clearing my recycle bin, but that didn't seem to help. In the user quotas, I do now have an entry under 'bytes' and under 'blocks' whereas [before it was 0 as seen in my previous attempts](https://github.com/afsc-gap-products/gapindex/issues/63#issuecomment-2617482498).

![Image](https://github.com/user-attachments/assets/4754055e-0d99-47f8-9a94-78cc0a8ef73e)

**Comment by zoyafuso-NOAA** [comment]
Right on. Okay, it makes sense that it could be a space issue, you do have a lower quota compared to the rest of us. You can use the stats that are on my computer as a guide as to how much to increase your max_bytes and max_blocks. 

We're so close.

**Comment by RebeccaHoward-NOAA** [comment]
Hi @zoyafuso-NOAA and @sowasser, jumping in to let you know that I'm having the same issue with one specific data pull but not others. Here's what I have as far as space goes:

![Image](https://github.com/user-attachments/assets/055c97d7-6d83-4335-a5f2-0ea073a4149d)

Did increasing max_bytes and max_blocks help with this and if so how do I go about doing that?

**Comment by zoyafuso-NOAA** [comment]
So I think the two solutions for this issue in this order is:

1) Check to make sure you have write access to your account/schema. If you don't, ask OFIS to update your account.
2) Check your Oracle account storage (using the query: SELECT * FROM USER_TS_QUOTAS) to make sure you have enough table space in your account. MAX_BYTES and MAX_BLOCKS are the important fields. If they are set to a low amount, also ask OFIS to upgrade your specs. Here are my current specs as a template. 

![Image](https://github.com/user-attachments/assets/d1d060aa-0c56-4694-8bda-6be56e75038c)
